MH-13471, Shibboleth SSO plugin to add roles for users on OC according to their EDUPERSONAFFILIATION. eg: "ROLE_AAI_USER_AFFILIATION_student" for "student"

aliyevTU opened the pull request:
Shibboleth SSO plugin to add roles for users on OC according to their EDUPERSONAFFILIATION.

Example 1: if Alice as EDUPERSONAFFILIATION has "student;" value then "ROLE_AAI_USER_AFFILIATION_student" will be added to OC for Alice.

Example 2: if Bob as EDUPERSONAFFILIATION has "student;member;staff;alum" value then
"ROLE_AAI_USER_AFFILIATION_student"
"ROLE_AAI_USER_AFFILIATION_member"
"ROLE_AAI_USER_AFFILIATION_staff"
"ROLE_AAI_USER_AFFILIATION_alum" roles will be added to OC for Bob.

Developed for the following initial ticket: https://opencast.jira.com/browse/MH-13223

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

smarquard commented:
Adding these as groups would provide a bit more flexibility than individual roles. e.g the Sakai provider will add the group ROLE_GROUP_SAKAI_INSTRUCTOR if you have instructor role in one or more sites.

So you could add

ROLE_GROUP_AAI_student
ROLE_GROUP_AAI_staff

which means you can use these groups in ACLs, but also give access to capabilities through the groups (e.g. differing Admin UI permissions, capture agent access, etc.)

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

imported java.util.Arrays

JulianKniephoff commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/ConfigurableLoginHandler.java:
```diff
@@ -425,6 +465,14 @@ private String extractEmail(HttpServletRequest request) {
+	if (headerAffiliation != null) {
+      String affiliation = request.getHeader(headerAffiliation);
+	  List<String> affiliations = Arrays.asList(affiliation.split(";"));
+	  for (String var : affiliations) {
```
`var` is such a generic name. Why is this not called `affiliation`?

aliyevTU replied:
I changed it as you suggested.

aliyevTU replied:
> 
So, that was why I named it "var". Changed it to "eachAffiliation".

`[ERROR] /home/travis/build/opencast/opencast/modules/security-aai/src/main/java/org/opencastproject/security/aai/ConfigurableLoginHandler.java:[471,23] variable affiliation is already defined in method extractRoles(java.lang.String,javax.servlet.http.HttpServletRequest)`

> I changed it as you suggested.

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

changed variable name "var" to "affiliation" in the for-loop as requested.

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

changed variable name "affiliation" to "eachAffiliation" in the for-loop, because of the variable name conflict.

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/ConfigurableLoginHandler.java:
```diff
@@ -252,6 +272,14 @@ public void updated(Dictionary properties) throws ConfigurationException {
       logger.warn("Optional header '{}' is not configured ", CFG_HEADER_HOME_ORGANIZATION_KEY);
     }
+	
+	String cfgAffiliation = StringUtils.trimToNull((String) properties.get(CFG_HEADER_AFFILIATION_KEY));
```
Indentation ;-)

aliyevTU replied:
Fixed

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/ConfigurableLoginHandler.java:
```diff
@@ -298,6 +326,18 @@ public void updated(Dictionary properties) throws ConfigurationException {
               CFG_ROLE_ORGANIZATION_SUFFIX_KEY, roleOrganizationSuffix);
+    }
+	
+	String cfgRoleAffiliationPrefix = StringUtils.trimToNull((String) properties.get(
```
Indentation

aliyevTU replied:
Thank you for bringing it to my attention. Fixed all the indentations :)

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

fixed indentation

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/ConfigurableLoginHandler.java:
```diff
@@ -425,6 +465,14 @@ private String extractEmail(HttpServletRequest request) {
       roles.add(new JpaRole(GLOBAL_ADMIN_ROLE, organization));
     }
+	if (headerAffiliation != null) {
+      String affiliation = request.getHeader(headerAffiliation);
```
Depending on the Shibboleth federation and the configuration of the Shibboleth service provider, it might be possible that this header is not set at all.

Therefore, you need to add the check "if (affiliation != null)".

aliyevTU replied:
Since this feature is optional, only if on the configuration file "header.affiliation" is uncommented and set to the correct value, this feature will work, therefore no check for "affiliation" value is necessary.

aliyevTU replied:
In case of a faulty configuration/typo by system administrator "affiliation" check added, as a safety measure.

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/ConfigurableLoginHandler.java:
```diff
@@ -425,6 +465,14 @@ private String extractEmail(HttpServletRequest request) {
+	if (headerAffiliation != null) {
+      String affiliation = request.getHeader(headerAffiliation);
+	  List<String> affiliations = Arrays.asList(affiliation.split(";"));
+	  for (String eachAffiliation : affiliations) {
```
No strong opinion, by eachAffiliation seems a somewhat weird choice for a variable name to me (e.g. read roleAffiliationPrefix + eachAffiliation). I suggest to use simply "affiliation" here.

aliyevTU replied:
Since there already exists a variable called "affiliation", and "var" is too generic, I settled with "eachAffiliation".

staubesv replied:
You could rename _affiliation_ to _affiliationHeader_, then the name _affiliation_ is free 

Anyway, no strong opinion here ;-)

staubesv commented on etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg:
```diff
@@ -51,3 +54,7 @@
 
+# Role prefix for affiliation
+# Default: ROLE_AAI_USER_AFFILIATION_
+#role.affiliation.prefix = ROLE_AAI_USER_AFFILIATION_
```
Maybe add a forth point after line https://github.com/opencast/opencast/pull/812/files#diff-6861a5481b6de717f8e1169f75581425R39

aliyevTU replied:
Thank you, done.

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION 

updated documentation on comments

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION 

added documentation about the new feature to the comments.

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION

added extra check as safety measure in case of a faulty/typo configuration.

aliyevTU committed:
MH-13471-add-roles-based-on-EDUPERSONAFFILIATION 

fixed whitespaces according to the style
